solr/core/src/test/org/apache/solr/store/blob/process/BlobDeleteProcessorTest.java

```diff
@@ -0,0 +1,472 @@
+    });
+    
+    // we shouldn't enqueue more than (numberOfTasks * defaultMaxAttempts) tasks to the pool 
+    assertTrue(enqueuedTasks.size() < (numberOfTasks * defaultMaxAttempts));
+    assertEquals(numberOfTasks, results.size());
+    int totalAttempts = 0;
+    for (BlobDeleterTaskResult res : results) {
+      assertNotNull(res);
+      assertNotNull(res.getTask());
+      assertEquals("scheduledFailingTask", res.getTask().getActionName());
+      totalAttempts += res.getTask().getAttempts();
+    }
+    // total task attempts should be consistent with our test scaffolding
+    assertTrue(totalAttempts < (numberOfTasks * defaultMaxAttempts));
+      
+    processor.shutdown();
+  }
+  
+  private List<BlobDeleterTask> generateRandomTasks(int defaultMaxAttempts, int taskCount) {
+    List<BlobDeleterTask> tasks = new LinkedList<>();
+    for (int i = 0; i < taskCount; i++) {
+      BlobDeleterTask task = null;
+      int totalAttempts = random().nextInt(defaultMaxAttempts);
+      int totalFails = random().nextInt(defaultMaxAttempts + 1);
+      task = buildScheduledFailingTaskForTest(blobClient, "test"+i, totalAttempts, true, totalFails);
+      tasks.add(task);
+    }
+    return tasks;
+  }
+  
+  /**
+   * Returns a test-only task for just holding onto a resource for test purposes
+   */
+  private BlobDeleterTask buildBlockingTaskForTest(CountDownLatch latch) {
+    return new BlobDeleterTask(null, null, false, 0) {
+      @Override
+      public Collection<String> doDelete() throws Exception {
+        // block until something forces this latch to count down
+        latch.await();
+        return null;
+      }
+      
+      @Override
+      public String getActionName() { return "blockingTask"; }
+    };
+  }
+  
+  /**
+   * Returns a test-only task that always fails on action execution by throwing an
+   * exception
+   */
+  private BlobDeleterTask buildFailingTaskForTest(CoreStorageClient client, 
+      String collectionName, int maxRetries, boolean allowRetries) {
+    return new BlobDeleterTask(client, collectionName, allowRetries, maxRetries) {
+      @Override
+      public Collection<String> doDelete() throws Exception {
+        throw new Exception("");
+      }
+      
+      @Override
+      public String getActionName() { return "failingTask"; }
+    };
+  }
+  
+  /**
+   * Returns a test-only task that fails a specified number of times before succeeding
+   */
+  private BlobDeleterTask buildScheduledFailingTaskForTest(CoreStorageClient client, 
+      String collectionName, int maxRetries, boolean allowRetries, int failTotal) {
+    return new BlobDeleterTask(client, collectionName, allowRetries, maxRetries) {
+      private AtomicInteger failCount = new AtomicInteger(0);
+      
+      @Override
+      public Collection<String> doDelete() throws Exception {
+        while (failCount.get() < failTotal) {
+          failCount.incrementAndGet();
+          throw new Exception("");
+        }
+        return null;
```

Comment:
The thread leak is due to the bug in the test which should be addressed now

Reply:
Even with test fixed, tests cleanup should always run (failure or no-failures). Otherwise, any future regression that causes a test to fail will start leaking threads. From past experience we know that, leaking threads negatively impacts the whole test run, not just the tests causing the leak.